lantiq: add support for AVM FRITZ!Box 3390

andyboeh opened the pull request:
The AVM FRITZ!Box 3390 is a lantiq-based device with two Atheros radios (2.4GHz and 5GHz). A few features make it special: the 5GHz WiFi is directly connected to the Lantiq CPU, but the 2.4GHz WiFi is connected to a separate ath79-based SoC. Both CPUs communicate via Ethernet.

This PR only adds support for the Lantiq side with the 5GHz WiFi. The 2.4GHz WiFi requires a separate ath79-based target along with some kernel patches and the necessary tools for uploading and configuring.

andyboeh committed:
lantiq: VR9: fix gpio-hog by defining the GPIO ranges

Signed-off-by: Andreas Böhler <dev@aboehler.at>

adrianschmutzler commented on target/linux/lantiq/base-files/etc/board.d/02_network:
```diff
@@ -143,7 +143,13 @@ avm,fritz3370-rev2-micron)
 	ucidef_add_switch "switch0" \
 		"0:lan:3" "1:lan:4" "2:lan:2" "4:lan:1" "6t@eth0"
 	;;
-
```
do not remove empty line.

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+	};
+
+	aliases {
+		led-boot = &power_green;
```
Please prefix labels by "led_", i.e. led_power_green, led_power_red

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+	leds {
+		compatible = "gpio-leds";
+
+		power_green: power {
```
Please prefix the _labels_ by led, and add the color to node names where not unique. In this case:
`led_power_green: power_green {`

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+			default-state = "keep";
+		};
+
+		power_red: power2 {
```
`led_power_red: power_red {`

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+			gpios = <&gpio 46 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi: wifi {
```
`led_wifi: wifi {`

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+			gpios = <&gpio 36 GPIO_ACTIVE_LOW>;
+		};
+
+		dsl: dsl {
```
`led_dsl: dsl {`

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+			gpios = <&gpio 47 GPIO_ACTIVE_LOW>;
+		};
+
+		info_green: info_green {
```
`led_info: info`

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+};
+
+&eth0 {
+	lan: interface@0 {
```
Is the lan label required?

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+		phy7: ethernet-phy@7 {
+			reg = <0x7>;
+			compatible = "avm,wasp", "ethernet-phy-ieee802.3-c22";
+			/* reset-gpios = <&gpio 34 GPIO_ACTIVE_HIGH>; */
```
What about this?

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/FRITZ3390.dts:
```diff
@@ -0,0 +1,352 @@
+	flash@4 {
+		compatible = "jedec,spi-nor";
+		reg = <4>;
+		spi-max-frequency = <1000000>;
```
1 Mhz?

adrianschmutzler commented on target/linux/lantiq/image/Makefile:
```diff
@@ -744,6 +744,17 @@ define Device/avm_fritz3370-rev2-micron
 endef
 TARGET_DEVICES += avm_fritz3370-rev2-micron
 
+define Device/avm_fritz3390
```
Note that there have been changes to lantiq recently:
- Device definitions have been moved to individual Makefiles per board.
- DTS files have been renamed by soc_vendor_model scheme.
- DEVICE_DTS is assembled automatically based on node name and to-be-specified SOC variable (in your case `SOC := vr9`)

adrianschmutzler commented on target/linux/lantiq/image/Makefile:
```diff
@@ -744,6 +744,17 @@ define Device/avm_fritz3370-rev2-micron
+define Device/avm_fritz3390
+  $(Device/AVM)
+  $(Device/NAND)
+  DEVICE_MODEL := FRITZ!Box 3390
```
This has been replaced by a newer syntax.

adrianschmutzler commented on target/linux/lantiq/image/Makefile:
```diff
@@ -744,6 +744,17 @@ define Device/avm_fritz3370-rev2-micron
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 49152k
+  DEVICE_DTS := FRITZ3390
+  DEVICE_PACKAGES := kmod-ath9k kmod-owl-loader wpad-basic kmod-usb-dwc2 fritz-tffs wasp_uploader
```
Please wrap to have max. 80 chars per line.

andyboeh commented:
I think I fixed all issues so far and removed the references to my personal packages. This PR adds, for now, only support for the 5GHz WiFi. 
The 2.4GHz WiFi requires another ath79-based target along with the necessary tools to upload and configure the firmware. I will create a separate PR for the ath79 target once this is merged.

blogic commented:
can you please rebase on latest changes in the tree so that we can merge this ?

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+		wifi {
+			label = "wlan";
+			gpios = <&gpio 29 GPIO_ACTIVE_HIGH>;
+			linux,code = <KEY_WLAN>;
```
Please use KEY_RFKILL unless you have a specific reason for not doing that.

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			urlader: partition@0 {
```
You do not seem to use that label in DTS, so you can remove it.

adrianschmutzler commented:
Hi, the adjustment for the base-files split should be relatively easy. Just move stuff for 01_leds, 12-ath9k-eeprom and platform.sh. For 02_network, you will need to split your code across the three functions, but that shouldn't be too hard either.

I'm personally not a big fan of adding kernel patches are mere diffs. It would be nice if you could format those as proper patch with explanation of what/why as in a commit message, though I will leave the final verdict on this to @blogic.

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+		compatible = "gpio-export";
+
+		gpio_wasp_reset {
+			gpio-export,name = "fritz3390:wasp:reset";
```
I assume that this is for the "other SoC"?

my personal on this topic:
I would work with the upstream linux maintainers to see how this can be integrated upstream (instead of going the route by getting it into OpenWrt without upstream backing). I have updated the lantiq target to newer kernel versions in the past and it was a pain. This is why I am upstreaming lantiq code whenver I have time (my latest mainline contribution for lantiq is the PCIe PHY driver in v5.4: ```drivers/phy/lantiq/phy-lantiq-vrx200-pcie.c```)

I don't know how communication with this "other SoC" works (I just read that it's some Atheros / QCA part).
However, there is a subsystem for "remote processors" in ```drivers/remoteproc/```
My high-level idea (again, without understanding the details - I think examples are better than long text) would look like this in .dts:
`
  &mdio {
    wasp: remote-processor@7 {
      comatible = "qca,wasp-remote-processor";
      reg = <0x7>;
      reset-gpios = <&gpio 34 GPIO_ACTIVE_HIGH>;
      firmware-name = "fritzbox3390-wasp-remote-processor-firmware"; /* optional */
    }
  };
`
(I don't know how sharing the configuration between the lantiq host processor and the QCA remote processor would work in that case. that's one of the reasons why I think it's important that this is discussed upstream. at least there are cases where ralink SoCs are connected similarly - so you could be the first to implement something that other drivers will re-use :))

andyboeh replied:
Yes, it's for the Atheros SoC. The communication works as follows:

  * Atheros is brought out of Reset by driving the reset GPIO high
  * a stage 1 firmware is uploaded via MDIO (called "ath_tgt_fw1.fw" in the original firmware); I'm uploading the original firmware for now
  * the stage 1 loader broadcasts ethernet frames and waits for the stage 2 upload which contains an actual Linux kernel.
  * the Atheros WiFi interface is exposed via the internal Gigabit connection to the Lantiq and looks like a native interface on the Lantiq. Basically, it looks a lot like an ath10k-htc implementation, but with a Gigabit connection instead of a USB connection. I went a slightly different approach here and uploaded a separate, complete OpenWrt initramfs image built for an ath79 target.

I implemented as much as possible in userspace, since the protocol for the upload is still a work in progress (I also started working on a 4490, it's very similar but requires some adaptations). In case you want to get a more thorough insight, the uploader code can be found at https://github.com/andyboeh/wasp_uploader.

andyboeh replied:
So I had a quick look into the remoteproc interface: To me, it seems as if it was meant for communicating with very specific cores or firmwares. While it would be suitable for supporting the WiFi functionality only, it would require a specific firmware to be written for the ath79 target (similar to the ath9k_htc firmware) - this is way over my head.

On the other hand, my approach leads to a fully customizable second OpenWrt instance on the same board which IMHO greatly improves the capability of the device.

xdarklight replied:
my understanding is that the remoteproc subsystem is used for booting "remote CPUs"
there's no definition how to communicate with the software (can be OpenWrt, some RTOS, DSP firmware, etc.).

but as you discovered: my idea is not perfect
that's why I suggested to discuss this with the maintainers of the Linux kernel (I think the remoteproc, general networking, wireless and MIPS subsystems are relevant in this discussion)

andyboeh replied:
I overslept it a few nights: Currently, I don't feel very comfortable with the firmware upload code. It's entirely reverse engineered, since I couldn't find any documentation on it. Thus, I would prefer a user-space solution that allows easier updates than an in-kernel solution. Plus, the 3490 uses a similar concept, but uses different registers which makes things more complex.

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+		};
+	};
+
+	pins_spi_default: pins_spi_default {
```
you can probably drop this since edb0a936f0af134ed2fb01c05e5d31b9944d2454

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pins_spi_default>;
```
like above, edb0a936f0af134ed2fb01c05e5d31b9944d2454 probably makes the pinctrl-* properties here obsolete

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+&pcie0 {
+	status = "okay";
+	gpio-reset = <&gpio 21 GPIO_ACTIVE_LOW>;
+	gpio-reset-dev = <&gpio 22 GPIO_ACTIVE_LOW>;
```
do you know what this gio-reset-dev pin is used for?
as far as I understand there is only one reset line in the PCIe connector spec, called WAKE# or PERST#: https://en.wikipedia.org/wiki/PCI_Express#Pinout

andyboeh replied:
I don't know what it is for, but if it is dropped, the PCIe bus does not respond at all. The same is true for the 3490 (there is an xhci controller on PCIe instead of WiFi). I found it in AVM kernel sources, they toggled it together with the reset pins (as the corresponding kernel patch now does).

xdarklight replied:
> In theory this means that we would have to add support for managing the reset-gpios inside the wifi@ node to the ath9k driver.
> But I don't know if that will work in reality, because I think (based on your description) the ath9k driver is only loaded after that reset line is deasserted (= AR9580 is taken out of reset).
I forgot to mention:
the USB subsystem has a similar problem and there's a workaround (called gpio-hog) which is accepted upstream, see here for an example: https://github.com/torvalds/linux/blob/6b14dd7267126931e9a95c68a442e8f2dabdc3c4/arch/arm/boot/dts/meson8b-odroidc1.dts#L285

andyboeh replied:
Good points, thanks for the research!
I have a similar problem on the 3490 as well and it seems that the ath79 target on the 3490 needs a reset line for the ath10k chip as well. 
Instead of the gpio-hog it might be sufficient to have a gpio-export for the reset line to force it high, I'll investigate that further and report back. It will take a few days, since I'm a bit busy with work.

andyboeh replied:
It works using gpio-hog, I removed the patch and added the gpio-hog to the .dts. Took me quite some time to figure out that gpio-ranges is required, otherwise it fails with -517 (see https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/080-pinctrl-msm-fix-gpio-hog-related-boot-issues.patch for the same problem on a different target). I fixed it for all vr9 boards.

Fun fact: my first intention was to simply set the GPIO to a predefined state, but since I wasn't aware of gpio-hog, I came up with the patch.

Adding support for the reset to ath9k wouldn't fix the issue: the PCIe link times out during initialization if the GPIO is not correctly set up, setting it in ath9k would be too late.
```
[    0.430881] ifx_pcie_wait_phy_link_up timeout
[    0.651469] ifx_pcie_wait_phy_link_up timeout
[    0.868731] ifx_pcie_wait_phy_link_up timeout
[    1.086316] ifx_pcie_wait_phy_link_up timeout
[    1.303914] ifx_pcie_wait_phy_link_up timeout
[    1.308149] pcie_rc_initialize link up failed!!!!!
```

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,338 @@
+		#address-cells = <2>;
+		device_type = "pci";
+
+		wifi@168c,0033 {
```
this should be wifi@0,0

I have done it wrong in the past (which is probably where you have copied this incorrect one from).
The pattern is: name@unit-address
unit-address has two-address-cells -> take the first two values from the "reg" property and separate them via a comma

adrianschmutzler replied:
@xdarklight OT: Is it still wrong there and can you correct it?

xdarklight replied:
I put this on my TODO-list (I may need a few weeks to take care of this though)

xdarklight commented on target/linux/lantiq/patches-4.19/2001-NET-MIPS-lantiq-add-mdio-ioctl.patch:
```diff
@@ -0,0 +1,43 @@
+ }
+ #endif
+ 
++static int xrx200_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
```
with this remoteprocessor driver idea from above you may be able to drop this ioctl

andyboeh replied:
Yes, depends on how much is going to be implemented in kernel space and how much in user space. 
Do you see disadvantages in allowing these ioctls? Other ethernet drivers expose the MDIO interface as well and I was quite surprised to find out the `mdio-tool` or `phytool` do not work on the Lantiq target.

xdarklight replied:
you're patching the out-of-tree driver here. by itself this is not a problem because that's what we use in OpenWrt.
my understanding is that the upstream driver (drivers/net/ethernet/lantiq_xrx200.c) in Linux 5.4 has the same feature set as the out-of-tree driver (before your patch), so switching to the upstream driver is a matter of writing a config migration script (to switch from the old to the new driver).

I don't know if the upstream driver also supports these IOCTL already. if not we may end up in a dilemma:
- use the upstream driver and break support for the FRITZ!Box 3390 (either by accident, because nobody knows that the IOCTLs are missing in the upstream driver - or on purpose, because it affects only one device)
- keep the out-of-tree driver and live with the increased maintenance effort

a slightly off-topic but still related story from the past: In the past we switched from the out-of-tree SPI driver to the upstream one during a kernel update because I could not get the out-of-tree one working (even after spending multiple days on it).

andyboeh replied:
I don't have a problem with submitting the patch to upstream, but I suppose that more work is needed before it can be accepted, since I only implemented the bare minimum of ioctls to support the FRITZ!Box. 

Since I'm also working on the 3490: It needs the same ioctls, so more than one device would be affected.

adrianschmutzler commented:
> > leave the final verdict on this to @blogic.
> 
> I think I got your point, you mean a header similar to the one in this patch: https://github.com/openwrt/openwrt/blob/master/target/linux/lantiq/patches-4.19/0005-lantiq_etop-pass-struct-device-to-DMA-API-functions.patch?

Yes, exactly. Format it like a real patch you would send to the kernel. This will make it easier for people just looking at the patch to see what it's for, instead of having to check the history and find the comment in the commit message here.

adrianschmutzler commented on target/linux/lantiq/xrx200/base-files/etc/hotplug.d/firmware/12-ath9k-eeprom:
```diff
@@ -19,6 +19,9 @@ case "$FIRMWARE" in
 			avm,fritz7360sl)
 				caldata_extract "urlader" 0x985 0x1000
 				;;
+			avm,fritz3390)
```
Please apply alphabetic sorting.

andyboeh replied:
I'm unsure how to sort this. The case for 3370 also contains 7362sl. Would you sort 3390 as the second case after 3370|7362sl, but before 7360sl?

adrianschmutzler replied:
Yes, the first member determines the sorting of the whole block.

adrianschmutzler commented on target/linux/lantiq/xrx200/base-files/etc/board.d/02_network:
```diff
@@ -130,7 +135,8 @@ lantiq_setup_macs()
 		;;
-	avm,fritz7362sl)
+	avm,fritz7362sl|\
+	avm,fritz3390)
```
Please apply alphabetic sorting.

andyboeh replied:
I'm not sure how to sort this. Just inside the case, i.e. swap 3390 and 7362sl, or both before 7360sl?

adrianschmutzler replied:
```
	avm,fritz3390|\
	avm,fritz7362sl)
		...
```
and then move the whole block up so it comes before fritz7360sl ...

adrianschmutzler commented:
Nitpick: Please remove the dot at the end of the commit title.

andyboeh commented:
For uploading the stage1 and stage2 firmware to the ath79 target, a separate tool is required (I called it "wasp_uploader"). Where should I submit the tool to (currently available at https://github.com/andyboeh/wasp_uploader).

Besides uploading the firmware files, it also generates a config.tar.gz file that includes the WiFi firmware and configuration files for the WiFi target. On the target, another tool, called "wasp_downloader" (currently available at https://github.com/andyboeh/wasp_downloader) is required for the configuration of the target system.

Ideally, it will be included in a default OpenWrt installation, so that the user just needs to put the WiFi target firmware (the separate ath79 OpenWrt installation I'll submit once the Lantiq patches have been accepted) into the correct location on the filesystem.

adrianschmutzler commented:
> (the separate ath79 OpenWrt installation I'll submit once the Lantiq patches have been accepted)

Note that an ath79 PR might receive more attention than a lantiq one, so consider submitting it in parallel.

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9.dtsi:
```diff
@@ -211,6 +211,7 @@
 			compatible = "lantiq,xrx200-pinctrl";
 			#gpio-cells = <2>;
 			gpio-controller;
+			gpio-ranges = <&gpio 0 0 56>;
```
Please add an explanation for this change to the commit message, or even make it a separate change, as it "affects" several devices.

andyboeh replied:
Good point, we should discuss that briefly: It could also be added to the FRITZ dts, I need it for the 3490 as well. When submitting the 3490 PR, I'll create a shared dts anyway, so that would still be an option. 
Without this definition, the gpio-hog doesn't work, thus I added it for all VR9 devices.

Where do you think does it belong to?

adrianschmutzler replied:
I cannot judge that from a technical point of view, but if you apply it for all vr9 devices, you 1. should make that clear and 2. should assess the impact on other devices at least punctually.

Maybe somebody with more substantial expertise on this target can help here.

mkresin replied:
> Without this definition, the gpio-hog doesn't work, thus I added it for all VR9 devices.

Done right for the correct reason. As requested by Adrian, please move this change into an own commit.

andyboeh replied:
Thanks, I created a separate commit.

adrianschmutzler commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,327 @@
+		};
+
+		led_wifi: wifi {
+			label = "fritz3390:green:wlan";
```
Change this to ":wifi" or change the node name/label to "wlan"/led_wlan.

andyboeh replied:
Whups, I missed that one...

adrianschmutzler commented on target/linux/lantiq/xrx200/base-files/etc/board.d/02_network:
```diff
@@ -43,6 +43,10 @@ lantiq_setup_interfaces()
 		;;
+	avm,fritz3390)
+		ucidef_add_switch "switch0" \
+			"0:lan:3" "1:lan:4" "2:lan:2" "4:lan:1" "5:lan:5" "6t@eth0"
```
Do you really have five LAN ports and no wan?

andyboeh replied:
Yes and now: Physically, there are only 4 LAN parts and no WAN, but one of the ports (number 4 IIRC) is the internal connection to the ath79 target. Thus, I considered it a LAN port. Should I add a comment?

adrianschmutzler replied:
I'm not sure whether this is the right way to do that, but again I'm lacking the technical expertise here, so I cannot give you a precise pointer.

Nevertheless, note that the number _after_ :lan: will be the port number displayed in Luci. So, if internal port no. 4 is just the internal connection, your external port numbers printed on the case should be 2-5 then, which I suspect is wrong.

andyboeh replied:
True, then it is probably port 5, the physical ports are attached to the same internal ports as for the 3370 model.

Since my current approach is to run a separate OpenWrt instance on the ath79 target, assigning it to LAN and mapping it to Port 5 by default makes sense to me.

andyboeh committed:
lantiq: add basic support for AVM FRITZ!Box 3390

The FRITZ!Box 3390 actually contains two SoCs, one Lantiq with a
5GHz WiFi and one AR9342 with a 2.4GHz WiFi. Only the Lantiq
has access to the flash memory, the Atheros runs fully from RAM.

Specifications
--------------

  - Lantiq 500 MHz
  - 128MiB RAM
  - 128MiB NAND
  - 256k Flash
  - AR9580 5GHz WiFi
  - AR9342 560 MHz
  - 64MiB RAM
  - AR9328 2.4GHz WiFi

Remarks
-------

This commit only adds support for the Lantiq side of things and
prepares the drivers for communication with the Atheros SoC. Thus,
only 5GHz WiFi works by default, the 2.4GHz WiFi will be added via
another target.

Some kernel patches will be required to add support for the Atheros SoC.

Installation
------------

Use the eva_ramboot.py script to boot the initramfs image. Then, transfer
the sysupgrade image to the device and run sysupgrade to flash it to the
NAND.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

andyboeh commented:
Just to be sure: is there anything left to do from my side, did I overlook something?

hvegh commented:
Is there anything we can do to get Fritz support moving to mainline?

xdarklight commented:
I didn't have time to review the current version yet (and I don't have enough time in the next days, but I can keep it on my TODO-list if you want me to).

There was a short discussion on IRC yesterday about the update to Linux 5.4.
We didn't make a final decision yet but there are plans to finally drop the out-of-tree Ethernet driver and use the DSA based on (which is part of the mainline kernel).

So my question to @andyboeh is:
how do you want to proceed with the Ethernet driver changes from this PR if we drop the out-of-tree Ethernet driver? the easiest way for us is to drop all code related to the out-of-tree Ethernet driver, but that will break functionality on the FRITZ!Box 3390.

andyboeh commented:
> how do you want to proceed with the Ethernet driver changes from this PR if we drop the out-of-tree Ethernet driver? the easiest way for us is to drop all code related to the out-of-tree Ethernet driver, but that will break functionality on the FRITZ!Box 3390.

The support from this PR does not really need the Ethernet driver patches, since they only prepare the support for the 2.4G WiFi module - which requires separate tools and another ath79 target as well, where we haven't decided what to do.

@xdarklight  My question is: do you think upstream would accept the two patches to the kernel? It's basically a handful of devices that require it, almost all from AVM.

xdarklight commented:
> > how do you want to proceed with the Ethernet driver changes from this PR if we drop the out-of-tree Ethernet driver? the easiest way for us is to drop all code related to the out-of-tree Ethernet driver, but that will break functionality on the FRITZ!Box 3390.
> 
> The support from this PR does not really need the Ethernet driver patches, since they only prepare the support for the 2.4G WiFi module - which requires separate tools and another ath79 target as well, where we haven't decided what to do.
I see, thank you for the summary
my suggestion is to split all the Ethernet code off into a separate pull request so basic support for the FRITZ!Box 3390 can be added with this PR.

> @xdarklight My question is: do you think upstream would accept the two patches to the kernel? It's basically a handful of devices that require it, almost all from AVM.
this is a tough question, here's my opinion:
- upstream will not accept the Ethernet IOCTL patches if their only benefit is to have some userspace program provide the firmware and then start "some remote processor"
- upstream will probably accept the Ethernet IOCTL patches if some good reason can be found (if any existing "standard" tool uses these IOCTLs then this may be a good reason to add support for them)
- upstream (I'm not sure exactly which subsystem, but I guess the linux-wireless, linux-netdev and possibly also the linux-mips and linux-remoteproc people) will be interested in discussing this with you. As far as I know you are the first person to make something like this work (regardless of whether the code is upstream or not). Some have tried because for example Ralink SoCs have a so-called iNIC feature which can can connect to the host CPU over Ethernet and do offloading similar to wasp on your board (AFAIK the Ralink code is closed-source). If you know how to make it run then the upstream developers will (at least from my experience) help you find the right driver subsystem(s) and help to develop a design/architecture that can be used for wasp and potentially for other "WiFi over Ethernet offloaders". And if you're lucky someone may help you out with some of the TODOs (so you only have to do half the work)

hvegh commented:
So my take is the following:

1. Merge basic support for 3390 (Without the provisions for loading the second target) from Andy

after this is complete I will:

2. rebase https://github.com/openwrt/openwrt/pull/2738 with basic support 7490 

and hopefully merge this after another round of review.

hvegh commented:
Sound like a plan ;)

andyboeh commented:
@hvegh @xdarklight I'll do my best, but usually I'm very limited with my spare time to work on this.

andyboeh commented:
So I removed the kernel patches, support for the offloading CPU will be added later.

@xdarklight The 5.4 testing kernel still uses the legacy xrx200 switch driver. Do you know if and when OpenWrt is going to switch to the upstream drivers?

xdarklight commented:
I reviewed the patches and they look good to me, they are:
Acked-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

> So I removed the kernel patches, support for the offloading CPU will be added later.
> 
> @xdarklight The 5.4 testing kernel still uses the legacy xrx200 switch driver. Do you know if and when OpenWrt is going to switch to the upstream drivers?

I am not sure but I have seen your message to openwrt-devel today.
I'll reply there

> I posted a message to the netdev mailing list, one suggestion was to create an MDIO driver which takes care of the stage 1 firmware upload. As far as I've understood the legacy driver, it probes the PHYs directly - without changes, this approach is not going to work.

meaning you need the DSA-based driver to make it work with the MDIO driver?

andyboeh commented:
I'm closing this, since I don't have much time to work on it and there is not much interest in the device. If somebody else wants to carry on, drop me a line.

jschwartzenberg commented:
Is this not closely related to #2738? I've been watching that PR and these two myself. I'm hoping that support for the x490 Fritzboxes will at least land as that's what I have to deal with for connectivity.

andyboeh commented:
> Is this not closely related to #2738? I've been watching that PR and these two myself. I'm hoping that support for the x490 Fritzboxes will at least land as that's what I have to deal with for connectivity.

Support for the 3490 can be found in my personal github as well. However, I'm not improving it due to the same reasons.

jschwartzenberg commented:
From what I understand looking through things very quickly, the 3490 is the same as the 7490 minus telephony interfaces. As I don't expect the 7490 support for OpenWRT to include telephony hardware support anyhow, those efforts seem effectively the same.

Maybe the efforts could be combined? The 7490 PR seems in quite a far stage already, but I don't know if your work could still improve it.

adrianschmutzler commented:
The problem is that several targets just lack a sufficient number of reviewers, or the people active on those targets are too limited in their time as well.
Unfortunately, there is not much you can do about that, except reviewing each others work. However, that will only lower the amount of work for the final reviewer, but it's still there.

jschwartzenberg commented:
I see. I imagine the reviews require quite deep knowledge of the kernel Linux and OpenWRT specifics themselves. Is there any way that vendors or ISPs could contribute? What about testing?

tmo26 commented:
@andyboeh Would be nice if you could include

- hardware info
- installation instructions

in your final PR.

DocMAX commented:
I pray to god that 3490 support will finally come to official openwrt...

andyboeh commented:
@tmo26 @DocMAX 
You do realize that this PR is closed and that I don't have any more interest in getting this merged? Feel free to carry on my work, in this case, please send me a message.

cocktailyogi commented:
Hi guys,

really, this is how it ends? I am also interested in 3390-support. The works seems to be done. What exactly is missing?
hardware info is in the PR and installation instructions are in the web as well.

Please grant that PR.

Kind regards, Yogi

hvegh commented:
I would be willing to put some testing and patching effort into it, but the
whole slugginess in the openwrt dev cycle regarding the AVM patches
basically instills me to run out of energy quickly.
So probably it ends here, unless somebody puts some time and effort into a
fork, which I consider unlikely.

Kind regards,
Henk

Op za 24 okt. 2020 om 09:52 schreef cocktailyogi <notifications@github.com>:

> Hi guys,
>
> really, this is how it ends? I am also interested in 3390-support. The
> works seems to be done. What exactly is missing?
> hardware info is in the PR and installation instructions are in the web as
> well.
>
> Please grant that PR.
>
> Kind regards, Yogi
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/2662#issuecomment-715876614>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCUNX3VTDWWNG3P5PCDZZLSMKBU3ANCNFSM4KBQ5HTA>
> .
>

cocktailyogi commented:
Hi people,
I was also interested in Fritzbox 3390 support. But I am not a professional developer. I did my best to help bringing in Andreas Böhlers work. So please excuse me, if I did something wrong.

I forked Andys work and rebased it to new Kernel of OpenWRT. And I have created an new PR to get it in. Unfortunately I have decided to remove Andys work on the 2nd SOC and WASP-part because I am not interested in 2.4GHz Wifi and it was too complicated to me. My priority was to make the Fritzbox 3390 a nice Router with VDSL modem and a managed 1Gbit Switch.

The new PR is here, maybe it can help somebody.

@andyboeh : Thx for your great work and as a beginner: Sorry if I did something wrong with your code. I am open for suggestions.

cocktailyogi commented:
@hvegh 

Are you still interested in helping? You could do a review for us:
https://github.com/openwrt/openwrt/pull/3656

Thx, Yogi
